frontend/drivers/platform_ctr.c

```diff
@@ -1,564 +1,584 @@
+}
+
+static enum frontend_powerstate frontend_ctr_get_powerstate(int* seconds, int* percent)
+{
+   u8 battery_percent = 0;
+   u8 charging = 0;
+   enum frontend_powerstate pwr_state = FRONTEND_POWERSTATE_NONE;
+
+   mcuHwcGetBatteryLevel(&battery_percent);
+   *percent = battery_percent;
+
+   /* 3ds does not support seconds of charge remaining */
+   *seconds = -1;
+
+   PTMU_GetBatteryChargeState(&charging);
+   if (charging)
+   {
+      if (battery_percent == 100)
+         pwr_state = FRONTEND_POWERSTATE_CHARGED;
+      else
+         pwr_state = FRONTEND_POWERSTATE_CHARGING;
+   }
+   else
+      pwr_state = FRONTEND_POWERSTATE_ON_POWER_SOURCE;
+
+   return pwr_state;
+}
+
+static void frontend_ctr_get_os(char* s, size_t len, int* major, int* minor)
+{
+   OS_VersionBin cver;
+   OS_VersionBin nver;
+
+   strlcpy(s, "3DS OS", len);
+   Result data_invalid = osGetSystemVersionData(&nver, &cver);
+   if (data_invalid == 0)
+   {
+      *major = cver.mainver;
+      *minor = cver.minor;
+   }
+   else
+   {
+      *major = 0;
+      *minor = 0;
+   }
+
+}
+
+static void frontend_ctr_get_name(char* s, size_t len)
+{
+   u8 device_model = 0xFF;
+   CFGU_GetSystemModel(&device_model);/*(0 = O3DS, 1 = O3DSXL, 2 = N3DS, 3 = 2DS, 4 = N3DSXL, 5 = N2DSXL)*/
+
+   switch (device_model)
+   {
+      case 0:
+         strlcpy(s, "Old 3DS", len);
+         break;
+      case 1:
+         strlcpy(s, "Old 3DS XL", len);
+         break;
+      case 2:
+         strlcpy(s, "New 3DS", len);
+         break;
+      case 3:
+         strlcpy(s, "Old 2DS", len);
+         break;
+      case 4:
+         strlcpy(s, "New 3DS XL", len);
+         break;
+      case 5:
+         strlcpy(s, "New 2DS XL", len);
+         break;
+
+      default:
+         strlcpy(s, "Unknown Device", len);
+         break;
+   }
+}
+
+frontend_ctx_driver_t frontend_ctx_ctr =
+{
+   frontend_ctr_get_environment_settings,
+   frontend_ctr_init,
+   frontend_ctr_deinit,
+   frontend_ctr_exitspawn,
+   NULL,                         /* process_args */
+   frontend_ctr_exec,
+#ifdef IS_SALAMANDER
+   NULL,
+#else
+   frontend_ctr_set_fork,
+#endif
+   frontend_ctr_shutdown,
+   frontend_ctr_get_name,
+   frontend_ctr_get_os,
+   frontend_ctr_get_rating,
+   NULL,                         /* load_content */
+   frontend_ctr_get_architecture,
+   frontend_ctr_get_powerstate,
+   frontend_ctr_parse_drive_list,
+   frontend_ctr_get_mem_total,
+   frontend_ctr_get_mem_used,
+   NULL,                         /* install_signal_handler */
+   NULL,                         /* get_signal_handler_state */
+   NULL,                         /* set_signal_handler_state */
+   NULL,                         /* destroy_signal_handler_state */
+   NULL,                         /* attach_console */
+   NULL,                         /* detach_console */
+   NULL,                         /* watch_path_for_changes */
+   NULL,                         /* check_for_path_changes */
+   "ctr",
+};

```

Comment:
what happened with this file ? windows line breaks ? makes it hard to review this way.